lightning/backend/local.go

```diff
@@ -892,65 +880,39 @@ WriteAndIngest:
 		endKey := codec.EncodeBytes([]byte{}, nextKey(pairEnd))
 		regions, err = paginateScanRegion(ctx, local.splitCli, startKey, endKey, 128)
 		if err != nil || len(regions) == 0 {
-			log.L().Warn("scan region failed", log.ShortError(err), zap.Int("region_len", len(regions)))
+			log.L().Warn("scan region failed", log.ShortError(err), zap.Int("region_len", len(regions)), zap.Int("retry", retry))
+			retry++
 			continue WriteAndIngest
 		}
 
-		shouldWait := false
-		errChan := make(chan error, len(regions))
 		for _, region := range regions {
 			log.L().Debug("get region", zap.Int("retry", retry), zap.Binary("startKey", startKey),
 				zap.Binary("endKey", endKey), zap.Uint64("id", region.Region.GetId()),
 				zap.Stringer("epoch", region.Region.GetRegionEpoch()), zap.Binary("start", region.Region.GetStartKey()),
 				zap.Binary("end", region.Region.GetEndKey()), zap.Reflect("peers", region.Region.GetPeers()))
 
-			// generate new uuid for concurrent write to tikv
-
-			if len(regions) == 1 {
-				w := local.ingestConcurrency.Apply()
-				rg, err1 := local.writeAndIngestPairs(ctx, engineFile, region, start, end)
-				local.ingestConcurrency.Recycle(w)
-				if err1 != nil {
-					err = err1
-					continue WriteAndIngest
-				}
-				if rg != nil {
-					remainRanges.add(*rg)
-				}
-			} else {
-				shouldWait = true
-				go func(r *split.RegionInfo) {
-					w := local.ingestConcurrency.Apply()
-					rg, err := local.writeAndIngestPairs(ctx, engineFile, r, start, end)
-					local.ingestConcurrency.Recycle(w)
-					errChan <- err
-					if err == nil && rg != nil {
-						remainRanges.add(*rg)
-					}
-				}(region)
-			}
-		}
-		if shouldWait {
-			shouldRetry := false
-			for i := 0; i < len(regions); i++ {
-				err1 := <-errChan
-				if err1 != nil {
-					err = err1
-					log.L().Warn("should retry this range", zap.Int("retry", retry), log.ShortError(err))
-					shouldRetry = true
+			w := local.ingestConcurrency.Apply()
+			var rg *Range
+			rg, err = local.writeAndIngestPairs(ctx, engineFile, region, pairStart, end)
+			local.ingestConcurrency.Recycle(w)
+			if err != nil {
+				regionRange := intersectRange(region.Region, Range{start: pairStart, end: end})
```

Comment:
could we only decode and compare two "start" 🤔

Reply:
done. PTAL, again